core/src/storage2/collections/binary_heap/elements.rs

```diff
@@ -0,0 +1,291 @@
+        Iter,
+        IterMut,
+        StorageVec,
+    },
+    traits::{
+        KeyPtr,
+        PackedLayout,
+        SpreadLayout,
+    },
+    Lazy,
+};
+
+/// Provides an interface for accessing elements in the `BinaryHeap`.
+///
+/// Elements of the heap are stored in a vector of `Children` objects, whereby
+/// each `Children` object contains two elements. When operating on indices of
+/// the `BinaryHeap` this interface transposes heap indices to the child inside
+/// the `Children` object, in which the element is stored.
+#[derive(Default, PartialEq, Eq, Debug)]
+pub struct Elements<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// The number of elements stored in the heap.
+    /// We cannot use the length of the storage vector, since each entry (i.e. each
+    /// `Children` object) in the vector contains two child elements (except the root
+    /// element which occupies a `Children` object on its own.
+    len: Lazy<u32>,
+    /// The underlying storage vec containing the `Children`.
+    children: StorageVec<Children<T>>,
+}
+
+impl<T> Elements<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// Creates a new empty storage heap.
+    pub fn new() -> Self {
+        Self {
+            len: Lazy::new(0),
+            children: StorageVec::new(),
+        }
+    }
+
+    /// Returns the number of elements in the heap, also referred to as its 'length'.
+    pub fn len(&self) -> u32 {
+        *self.len
+    }
+
+    /// Returns the amount of `Children` objects stored in the vector.
+    #[allow(dead_code)]
+    #[cfg(all(test, feature = "ink-fuzz-tests"))]
+    pub fn children_count(&self) -> u32 {
+        self.children.len()
+    }
+
+    /// Returns `true` if the heap contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns a shared reference to the indexed element.
+    ///
+    /// Returns `None` if `index` is out of bounds.
+    pub fn get(&self, index: u32) -> Option<&T> {
+        let storage_index = children::get_children_storage_index(index);
+        let i = self.within_bounds(storage_index)?;
+        let children = self.children.get(i)?;
+        let child_pos = children::get_child_pos(index);
+        children.child(child_pos)
+    }
+
+    /// Returns an exclusive reference to the indexed element.
+    /// The element in a `Children` is an `Option<T>`.
+    ///
+    /// Returns `None` if `index` is out of bounds.
+    pub fn get_mut(&mut self, index: u32) -> Option<&mut Option<T>> {
+        let storage_index = children::get_children_storage_index(index);
+        self.within_bounds(storage_index)?;
+        self.children.get_mut(storage_index).map(|children| {
+            let child_pos = children::get_child_pos(index);
+            children.child_mut(child_pos)
+        })
+    }
+
+    /// Swaps the elements at the given indices.
+    ///
+    /// # Panics
+    ///
+    /// If one or both indices are out of bounds.
+    pub fn swap(&mut self, a: u32, b: u32) {
```

Comment:
Also is this API required for a binary heap implementation? I do not see a need for it when looking at Rust's `BinaryHeap` implementation. https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html